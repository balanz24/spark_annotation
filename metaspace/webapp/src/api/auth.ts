const fetchPostJson = async(url: string, body: object) =>
  fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'include',
    body: JSON.stringify(body),
  })

export const signOut = async() => {
  const response = await fetch('/api_auth/signout', { method: 'POST', credentials: 'include' })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const signInByEmail = async(email: string, password: string): Promise<boolean> => {
  const response = await fetchPostJson('/api_auth/signin', { email, password })
  return response.status >= 200 && response.status < 300
}

export const createAccountByEmail = async(email: string, password: string, name: string) => {
  const response = await fetchPostJson('/api_auth/createaccount', { email, password, name })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const sendPasswordResetToken = async(email: string) => {
  const response = await fetchPostJson('/api_auth/sendpasswordresettoken', { email })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}

export const validatePasswordResetToken = async(token: string, email: string): Promise<boolean> => {
  const response = await fetchPostJson('/api_auth/validatepasswordresettoken', { token, email })
  return response.status === 200
}

export const resetPassword = async(token: string, email: string, password: string) => {
  const response = await fetchPostJson('/api_auth/resetpassword', { token, email, password })
  if (response.status !== 200) {
    throw new Error(`Unexpected response from server: ${response.status} ${response.statusText}`)
  }
}
