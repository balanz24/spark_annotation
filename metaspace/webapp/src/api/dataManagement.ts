import gql from 'graphql-tag'

export interface GroupListItem {
  id: string;
  name: string;
}

export const oneGroupQuery =
  gql`query oneGroupQuery($groupId: ID!) {
  group(groupId:$groupId) {
    id
    name
  }
}`

export const oneProjectQuery =
  gql`query oneProjectQuery($projectId: ID!) {
  project(projectId:$projectId) {
    id
    name
  }
}`

export const allGroupsQuery =
gql`query allGroupsQuery($query: String) {
  allGroups(query:$query) {
    id
    name
  }
}`

export const requestAccessToGroupMutation =
gql`mutation($groupId: ID!) {
  requestAccessToGroup(groupId: $groupId){
    role
    }
  }`
