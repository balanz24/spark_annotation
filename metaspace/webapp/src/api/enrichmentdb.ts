import gql from 'graphql-tag'

export interface EnrichmentDB {
  id: number
  name: string
}

export interface EnrichmentTerm {
  id: number
  enrichmentId: string
  enrichmentName: string
  enrichmentDB: EnrichmentDB
}

export const checkIfEnrichmentRequested =
gql`query enrichmentRequested($id: String!) {
  enrichmentRequested(datasetId: $id)
}`

export const getEnrichedMolDatabasesQuery =
gql`query allEnrichedMolDatabases($id: String!) {
  allEnrichedMolDatabases(datasetId: $id) {
    id
    name
    version
    archived
    group {
      id
      shortName
    }
  }
}`
