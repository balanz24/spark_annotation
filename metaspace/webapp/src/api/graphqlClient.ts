import { ApolloClient } from 'apollo-client-preset'
import { BatchHttpLink } from 'apollo-link-batch-http'
import { WebSocketLink } from 'apollo-link-ws'
import { setContext } from 'apollo-link-context'
import { SubscriptionClient } from 'subscriptions-transport-ws'
import { getOperationAST } from 'graphql/utilities/getOperationAST'
import { onError } from 'apollo-link-error'

import config from '../lib/config'
import tokenAutorefresh from './tokenAutorefresh'
import reportError from '../lib/reportError'
import { get } from 'lodash-es'
import { makeApolloCache } from '../lib/apolloCache'

const graphqlUrl = config.graphqlUrl || `${window.location.origin}/graphql`
const wsGraphqlUrl = config.wsGraphqlUrl || `${window.location.origin.replace(/^http/, 'ws')}/ws`

let $alert: ((message: string, title: string, options?: any) => Promise<any>) | null = null

export function setMaintenanceMessageHandler(_$alert: (message: string, title: string, options?: any) => Promise<any>) {
  $alert = _$alert
}

const isReadOnlyError = (error: any) => {
  try {
    return JSON.parse(error.message).type === 'read_only_mode'
  } catch {
    return false
  }
}

const authLink = setContext(async() => {
  try {
    return ({
      headers: {
        authorization: `Bearer ${await tokenAutorefresh.getJwt()}`,
      },
    })
  } catch (err) {
    reportError(err)
    throw err
  }
})

const errorLink = onError(({ graphQLErrors, networkError, forward, operation }) => {
  if (graphQLErrors) {
    const readOnlyErrors = graphQLErrors.filter(isReadOnlyError)

    if (readOnlyErrors.length > 0) {
      if ($alert != null) {
        readOnlyErrors.forEach(err => { (err as any).isHandled = true })
        $alert('This operation could not be completed. '
          + 'METASPACE is currently in read-only mode for scheduled maintenance. Please try again later.',
        'Scheduled Maintenance',
        { type: 'error' })
          .catch(() => { /* Ignore exception raised when alert is closed */ })
      }
    }
  } else if (networkError) {
    const message = get(networkError, 'result.message')
    if (message === 'jwt expired') {
      // noinspection JSIgnoredPromiseFromCall
      tokenAutorefresh.refreshJwt(true)
      return forward(operation)
    }
  }
})

const httpLink = new BatchHttpLink({
  uri: graphqlUrl,
  batchInterval: 10,
})

const wsClient = new SubscriptionClient(wsGraphqlUrl, {
  reconnect: true,
  async connectionParams() {
    // WORKAROUND: This is not the right place for this, but it's the only callback that gets called after a reconnect
    // and can run an async operation before any messages are sent.
    // All subscription operations need to have their JWTs updated before they are reconnected, so do that before
    // supplying the connection params.
    const operations = Object.values(wsClient.operations || {}).map(op => op.options)
    // @ts-ignore the private unsentMessagesQueue
    const queuedMessages: any[] = wsClient.unsentMessagesQueue.map((m: any) => m.payload)
    const payloads = [...operations, ...queuedMessages]

    if (payloads.length > 0) {
      const jwt = await tokenAutorefresh.getJwt()
      payloads.forEach(payload => {
        payload.jwt = jwt
      })
    }

    return {}
  },
})
wsClient.use([{
  async applyMiddleware(operationOptions: any, next: Function) {
    // Attach a JWT to each request
    try {
      operationOptions.jwt = await tokenAutorefresh.getJwt()
    } catch (err) {
      reportError(err)
      next(err)
    } finally {
      next()
    }
  },
}])

const wsLink = new WebSocketLink(wsClient)

const link = errorLink.concat(authLink).split(
  (operation) => {
    // Only send subscriptions over websockets
    const operationAST = getOperationAST(operation.query, operation.operationName)
    return operationAST != null && operationAST.operation === 'subscription'
  },
  wsLink,
  httpLink,
)

const apolloClient = new ApolloClient({
  link,
  cache: makeApolloCache(),
  defaultOptions: {
    query: {
      fetchPolicy: 'network-only',
    },
  },
})

export const refreshLoginStatus = async() => {
  // Problem: `refreshJwt` updates the Vuex store, which sometimes immediately triggers new queries.
  // `apolloClient.resetStore()` has an error if there are any in-flight queries, so it's not suitable to run it
  // immediately after `refreshJwt`.
  // Solution: Split the `resetStore` into two parts: invalidate old data before `refreshJwt` updates Vuex,
  // then ensure that all queries are refetched.

  await apolloClient.queryManager.clearStore()
  await tokenAutorefresh.refreshJwt(true)

  try {
    await apolloClient.reFetchObservableQueries()
  } catch (err) {
    // reFetchObservableQueries throws an error if any queries fail.
    // Let the source of the query handle it instead of breaking `refreshLoginStatus`.
    console.error(err)
  }
}

export default apolloClient
