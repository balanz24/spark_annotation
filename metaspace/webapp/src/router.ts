/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
