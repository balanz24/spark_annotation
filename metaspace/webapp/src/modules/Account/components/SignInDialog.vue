<template>
  <el-dialog
    title="Sign in"
    visible
    :lock-scroll="false"
    @close="onClose"
  >
    <div class="container">
      <div class="left-side">
        <el-form
          ref="form"
          :model="model"
          :rules="rules"
        >
          <el-form-item
            prop="email"
            :validate-status="isInvalid ? 'error' : undefined"
          >
            <el-input
              v-model="model.email"
              placeholder="Email address"
            />
          </el-form-item>
          <el-form-item
            prop="password"
            :validate-status="isInvalid ? 'error' : undefined"
          >
            <el-input
              v-model="model.password"
              type="password"
              placeholder="Password"
              @keypress.native.enter="onSubmit"
            />
          </el-form-item>
          <vue-slide-up-down :active="isInvalid">
            <p style="margin-top: 0; color: red">
              Incorrect username or password
            </p>
          </vue-slide-up-down>
          <el-button
            type="primary"
            :loading="isSubmitting"
            @click="onSubmit"
          >
            Sign in
          </el-button>
        </el-form>
      </div>
      <div class="mid">
        <div class="divider" />
        <div class="text">
          or
        </div>
        <div class="divider" />
      </div>
      <div class="right-side">
        <a
          class="google-button"
          href="/api_auth/google"
          @click="setSignInReturnUrl"
        >
          <google-button>Sign in with Google</google-button>
        </a>
      </div>
    </div>
    <p>
      <inter-dialog-link dialog="forgotPassword">
        Forgot password?
      </inter-dialog-link>
    </p>
    <p>
      New user? <inter-dialog-link dialog="createAccount">
        Create an account
      </inter-dialog-link>
    </p>
    <p style="margin-bottom: 0">
      Submitted data previously but never set a password? <a
        href="#"
        @click.prevent="togglePreMigrationUserInfo"
      >Click here</a>
    </p>
    <vue-slide-up-down :active="showPreMigrationUserInfo">
      <p style="margin-bottom: 0;">
        We now require that all METASPACE users have an account to upload and manage datasets.
        If you have submitted datasets to METASPACE previously,
        please <inter-dialog-link dialog="createAccount">
          create an account
        </inter-dialog-link> using the same email address
        that you used when submitting the datasets, and you will be reunited with them.
        <a href="mailto:contact@metaspace2020.eu">Contact us</a> if you have lost access to any datasets that you submitted.
      </p>
    </vue-slide-up-down>
  </el-dialog>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Watch } from 'vue-property-decorator'
import VueSlideUpDown from 'vue-slide-up-down'
import { Form } from '../../../lib/element-ui'
import GoogleButton from './GoogleButton.vue'
import InterDialogLink from './InterDialogLink'
import { signInByEmail } from '../../../api/auth'
import { refreshLoginStatus } from '../../../api/graphqlClient'
import reportError from '../../../lib/reportError'
import { setSignInReturnUrl } from '../signInReturnUrl'

  interface Model {
    email: string;
    password: string;
  }

  @Component({
    components: {
      VueSlideUpDown,
      GoogleButton,
      InterDialogLink,
    },
  })
export default class SignInDialog extends Vue {
    isSubmitting: boolean = false;
    isInvalid: boolean = false;
    showPreMigrationUserInfo: boolean = false;
    model: Model = {
      email: '',
      password: '',
    };

    rules = {
      email: [{ required: true, message: 'Email address is required' }],
      password: [{ required: true, message: 'Password is required' }],
    };

    @Watch('model', { deep: true })
    onModelChanged() {
      this.isInvalid = false
    }

    togglePreMigrationUserInfo() {
      this.showPreMigrationUserInfo = !this.showPreMigrationUserInfo
    }

    async onSubmit() {
      const form = this.$refs.form as Form
      await form.validate()
      try {
        const { email, password } = this.model
        this.isSubmitting = true
        const authenticated = await signInByEmail(email, password)
        if (authenticated) {
          await refreshLoginStatus()
          this.$store.commit('account/hideDialog', { dialog: 'signIn', isLoginSuccess: true })
        } else {
          form.clearValidate()
          this.isInvalid = true
        }
      } catch (err) {
        reportError(err)
      } finally {
        this.isSubmitting = false
      }
    }

    setSignInReturnUrl() {
      setSignInReturnUrl(this.$route)
    }

    onClose() {
      this.$store.commit('account/hideDialog', 'signIn')
    }
}
</script>

<style scoped lang="scss">
  @import "~element-ui/packages/theme-chalk/src/mixins/mixins";

  .el-dialog__wrapper /deep/ .el-dialog {
    width: 400px;
  }

  .container {
    display: flex;
    flex-direction: column;
    align-items: stretch;
  }

  .left-side {
    flex: 0 1 auto;
  }

  .right-side {
    flex: 0 1 auto;
  }

  .mid {
    flex: 0 0 50px;
    display: flex;
    flex-direction: row;
    align-items: center;

    .divider {
      flex-grow: 1;
      border-top: 1px solid lightgrey;
    }
    .text {
      flex: none;
      padding: 0.5em;
    }
  }

  @include res('sm') {
    .el-dialog__wrapper /deep/ .el-dialog {
      width: 690px;
    }
    .container {
      flex-direction: row;
    }
    .left-side {
      width: 360px;
    }
    .right-side {
      width: 240px;
    }
    .mid {
      flex-direction: column;
      .divider {
        border-top: none;
        border-left: 1px solid lightgrey;
      }
    }
  }

  .el-button {
    width: 100%;
  }

  .google-button {
    text-decoration: none;
  }
</style>
