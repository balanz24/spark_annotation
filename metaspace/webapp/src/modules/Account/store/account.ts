import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
