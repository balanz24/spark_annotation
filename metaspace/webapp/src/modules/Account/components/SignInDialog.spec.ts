import { mount, Wrapper } from '@vue/test-utils'
import { Button, FormItem } from '../../../lib/element-ui'
import Vuex from 'vuex'
import Vue from 'vue'
import SignInDialog from './SignInDialog.vue'
import account from '../store/account'
import router from '../../../router'
import { restoreConsole, suppressConsoleWarn } from '../../../../tests/utils/suppressConsole'
import * as _mockAuthApi from '../../../api/auth'
import { refreshLoginStatus as _mockRefreshLoginStatus } from '../../../api/graphqlClient'

jest.mock('../../../api/auth')
const mockAuthApi = _mockAuthApi as jest.Mocked<typeof _mockAuthApi>

jest.mock('../../../api/graphqlClient')
const mockRefreshLoginStatus = _mockRefreshLoginStatus as jest.Mocked<typeof _mockRefreshLoginStatus>

Vue.use(Vuex)

const setFormField = (wrapper: Wrapper<Vue>, fieldName: string, value: string) => {
  wrapper
    .findAllComponents<FormItem>(FormItem)
    .filter((fi: Wrapper<FormItem>) => fi.props().prop === fieldName)
    .at(0)
    .find('input')
    .setValue(value)
}

describe('SignInDialog', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    suppressConsoleWarn('async-validator:')
  })

  afterEach(() => {
    restoreConsole()
  })

  const store = new Vuex.Store({
    modules: {
      account: account,
    },
  })

  it('should match snapshot', () => {
    const wrapper = mount(SignInDialog, { store, router })
    expect(wrapper).toMatchSnapshot()
  })

  it('should be able to sign in', async() => {
    // Arrange
    const email = 'test@example.com'
    const password = 'baz'
    const wrapper = mount(SignInDialog, { store, router }) as Wrapper<SignInDialog>
    mockAuthApi.signInByEmail.mockImplementation(() => Promise.resolve(true))
    store.commit('account/showDialog', 'signIn')

    // Act
    setFormField(wrapper, 'email', email)
    setFormField(wrapper, 'password', password)
    wrapper.findComponent(Button).trigger('click')
    await Vue.nextTick()

    // Assert
    expect(mockAuthApi.signInByEmail).toBeCalledWith(email, password)
    expect(mockRefreshLoginStatus).toBeCalled()
    expect(store.state.account.dialog).toBe(null)
  })
})
