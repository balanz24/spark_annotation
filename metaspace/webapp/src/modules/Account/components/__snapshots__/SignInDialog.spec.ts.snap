// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInDialog should match snapshot 1`] = `
<mock-el-dialog title="Sign in" visible="">
  <div class="container">
    <div class="left-side">
      <form class="el-form">
        <div class="el-form-item is-required">
          <!---->
          <div class="el-form-item__content">
            <div class="el-input">
              <!----><input type="text" autocomplete="off" placeholder="Email address" class="el-input__inner">
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <transition-stub name="el-zoom-in-top">
              <!---->
            </transition-stub>
          </div>
        </div>
        <div class="el-form-item is-required">
          <!---->
          <div class="el-form-item__content">
            <div class="el-input">
              <!----><input type="password" autocomplete="off" placeholder="Password" class="el-input__inner">
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <transition-stub name="el-zoom-in-top">
              <!---->
            </transition-stub>
          </div>
        </div>
        <div style="overflow: hidden; transition-property: height; height: auto; transition-duration: 500ms;">
          <p style="margin-top: 0px; color: red;">
            Incorrect username or password
          </p>
        </div> <button type="button" class="el-button el-button--primary">
          <!---->
          <!----><span>
          Sign in
        </span></button>
      </form>
    </div>
    <div class="mid">
      <div class="divider"></div>
      <div class="text">
        or
      </div>
      <div class="divider"></div>
    </div>
    <div class="right-side"><a href="/api_auth/google" class="google-button">
        <div class="abcRioButton abcRioButtonBlue">
          <div class="abcRioButtonContentWrapper">
            <div class="abcRioButtonIcon" style="padding: 10px;">
              <div class="abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18" style="width: 18px; height: 18px;"><svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="18px" height="18px" viewBox="0 0 48 48" class="abcRioButtonSvg">
                  <g>
                    <path fill="#EA4335" d="M24 9.5c3.54 0 6.71 1.22 9.21 3.6l6.85-6.85C35.9 2.38 30.47 0 24 0 14.62 0 6.51 5.38 2.56 13.22l7.98 6.19C12.43 13.72 17.74 9.5 24 9.5z"></path>
                    <path fill="#4285F4" d="M46.98 24.55c0-1.57-.15-3.09-.38-4.55H24v9.02h12.94c-.58 2.96-2.26 5.48-4.78 7.18l7.73 6c4.51-4.18 7.09-10.36 7.09-17.65z"></path>
                    <path fill="#FBBC05" d="M10.53 28.59c-.48-1.45-.76-2.99-.76-4.59s.27-3.14.76-4.59l-7.98-6.19C.92 16.46 0 20.12 0 24c0 3.88.92 7.54 2.56 10.78l7.97-6.19z"></path>
                    <path fill="#34A853" d="M24 48c6.48 0 11.93-2.13 15.89-5.81l-7.73-6c-2.15 1.45-4.92 2.3-8.16 2.3-6.26 0-11.57-4.22-13.47-9.91l-7.98 6.19C6.51 42.62 14.62 48 24 48z"></path>
                    <path fill="none" d="M0 0h48v48H0z"></path>
                  </g>
                </svg></div>
            </div> <span class="abcRioButtonContents" style="font-size: 16px; line-height: 38px;">Sign in with Google</span>
          </div>
        </div>
      </a></div>
  </div>
  <p><a href="/account/forgot-password">
      Forgot password?
    </a></p>
  <p>
    New user? <a href="/account/create-account">
      Create an account
    </a></p>
  <p style="margin-bottom: 0px;">
    Submitted data previously but never set a password? <a href="#">Click here</a></p>
  <div style="overflow: hidden; transition-property: height; height: auto; transition-duration: 500ms;">
    <p style="margin-bottom: 0px;">
      We now require that all METASPACE users have an account to upload and manage datasets.
      If you have submitted datasets to METASPACE previously,
      please <a href="/account/create-account">
        create an account
      </a> using the same email address
      that you used when submitting the datasets, and you will be reunited with them.
      <a href="mailto:contact@metaspace2020.eu">Contact us</a> if you have lost access to any datasets that you submitted.
    </p>
  </div>
</mock-el-dialog>
`;
